### ISSUE TITLE
[Discussion] Proposed layer reorganization

### ISSUE THREAD

**Issue by ian-r-rose** [dask/dask#7722]
High level graph `Layer` implementations have a number of constraints on them which makes their design and location within the project somewhat subtle. For security reasons they cannot deserialize or otherwise run arbitrary python code, and in general they cannot depend on numpy, pandas, or other such complex dependencies. As a result of the latter requirement, many `Layer` implementations have been  placed in `dask.layers` and, where they can be safely referenced without triggering unwanted imports (cf #7381).

During some discussions, @gjoseph92, @GenevieveBuckley and I have become concerned about the size and maintainability of `layers.py`. As we look towards implementing more layers, this file will get more unwieldy: each layer will typically have an implementation and possibly a set of utility functions, and the module will eventually become a significant fraction of dask.

Since the natural location for these layers (near userland code that instantiates them) isn't workable due to the unwanted imports, we have sketched out a proposed structure for making this easier to organize and maintain going forward. The basic idea is to keep shadow submodules of `dask.array` and `dask.dataframe` (and possibly `dask.bag`) as siblings of those submodules, each containing layer implementations for their counterpart. So `dask.array_layers` would have `Layer` classes and all the supporting utilities for `dask.array`. This could look something like (conceptually, not literally):

```
├── core.py
├── base.py
├── array
      ├── overlap.py (user code for overlap)
      └──  slicing.py (user code for slicing)
├── dataframe
      ├── shuffle.py (user code for shuffles)
      └── io.py (user code for read_csv, read_parquet, etc)
├── array_layers
      ├── overlap.py (layers for overlap)
      └──  slicing.py (layers for slicing)
└── dataframe_layers
      ├── shuffle.py (layers for shuffles)
      └── io.py (layers for read_csv, read_parquet, etc)
```

Code from `dask.array` could freely import code from `dask.array_layers`, but not the reverse. I don't think the layers submodules would exactly mirror their counterparts, but a rough correspondence for collection-producing operations would make it much easier to keep things organized and avoid having `layers.py` be an enormous dumping ground.

Thoughts? especially cc @rjzamora and @jrbourbeau

**Comment by rjzamora** [comment]
Thanks for writing this up @ian-r-rose !

There is no doubt that `layers.py` needs to be reorganized into a more-maintainable module structure.  The only reason we didn't do it immediately was so we could have a larger discussion like this.

Do you have a specific reason to prefer the shadow `*_layers` structure over an organization where there is a dedicated `layers/` directory with `array` and `dataframe` submodules?  My only worry is that there may be significant layer logic that can/should be shared between the dataframe and array APIs.  Also, I'm curious if you have thoughts on *where* Blockwise-based layer logic should live moving forward.  Right now, some layer-related definitions (like `BlockwiseDep`) live in `blockwise.py`, because that is another "import-safe" module.  Do you think we should keep `Blockwise` in this module, and move everything else somewhere more "layer-specific"?

**Comment by gjoseph92** [comment]
> a dedicated layers/ directory with `array` and `dataframe` submodules?

Based on my experience building sheds for bicycles, I'd also prefer this structure. I guess it feels a little more equivalent to the `array`/`dataframe` submodules for them to have exactly the same names. But mostly, I think it will be nice to have a `dask/layers` base module for exactly the sort of

> significant layer logic that can/should be shared

that we'll inevitably end up with. I'd probably vote for moving `blockwise.py`, and maybe even `highlevel.py` (though maybe there are issues with that one I haven't thought of), into `dask/layers` as well.

**Comment by GenevieveBuckley** [comment]
I'd second Gabe's last comment - this feels like an organization structure I'd be able to wrap my head around relatively easily.

 (Mind you, I'm new to this area of the code in general, so I don't have any special insight on avoiding potential problems down the track)

**Comment by GenevieveBuckley** [comment]
Also, Ian suggested we might want a `dask.layer_utils` - see this for context https://github.com/dask/dask/pull/7595#discussion_r641015828

**Comment by mrocklin** [comment]
Crazy idea, what if we left these in dask/dataframe or dask/array and just
accepted the added requirements on the scheduler.

This is somewhat unclean, but we might do it anyway and see if anyone
complains.

In practice everyone I know deploys the same software environment on the
scheduler as they do on the workers.

I've always been the one to argue for this cleanliness, but I wanted us to
be aware of breaking this cleanliness as an option.

**Comment by jrbourbeau** [comment]
Thanks for getting this conversation started @ian-r-rose!

FWIW, like @rjzamora and @gjoseph92, I would prefer a single new `layers/` directory over multiple `*_layers/` directories for each collection type.

That said, I'd like to propose another option where we stick with our existing directory structure by having separate `array/layers.py`, `dataframe/layers.py`, etc. files for the various Dask collections, as well as a root `layers.py` file for `Layer` code which is used in multiple collections. To me this seems like a natural approach as we already do this today with files like `core.py` and `utils.py`. 

Re: allowing the scheduler to run NumPy, pandas, ... code. That's certainly something we should consider. I'm also not aware of any deployments where the scheduler and workers have separate software environments. Do folks who are writing `Layer` classes feel pain around not being able to use libraries like NumPy and pandas?

**Comment by rjzamora** [comment]
>I'd like to propose another option where we stick with our existing directory structure by having separate array/layers.py, dataframe/layers.py, etc. files for the various Dask collections, as well as a root layers.py file for Layer code which is used in multiple collections.

I don't think I follwo this. Are you suggesting that we add `layers.py` files to the existing `dask.array` and `dask.dataframe` modules?

>Do folks who are writing Layer classes feel pain around not being able to use libraries like NumPy and pandas?

I do think there are cases where it would be very convenient to bake numpy/pandas-API usge into the graph-materialization logic.  For example, when slicing out chunks of an Array/DataFrame for each task that is generated.

**Comment by jrbourbeau** [comment]
> Are you suggesting that we add layers.py files to the existing dask.array and dask.dataframe modules?

Yep, that's exactly it. This will help avoid a single, unwieldy `layers.py` file as we add more `Layer`s (as Ian  mentioned in his original comment), but will also conform nicely to our existing file structure. Specifically, I'm proposing something like:


```
dask/
├── ...
├── layers.py (layers shared across multiple collections go here)
├── array/
      ├── ...
      └── layers.py (array layers go here)
├── dataframe/
      ├── ...
      └── layers.py (dataframe layers go here)
├── ...
```

**Comment by ian-r-rose** [comment]
> That said, I'd like to propose another option where we stick with our existing directory structure by having separate `array/layers.py`, `dataframe/layers.py`, etc. files for the various Dask collections, as well as a root `layers.py` file for `Layer` code which is used in multiple collections. To me this seems like a natural approach as we already do this today with files like `core.py` and `utils.py`. Do others have thoughts on this structure?

This is certainly cleaner, and we discussed it previously in #7381. I just want to flag that a consequence of this is that you will need numpy/pandas installed in order to import/instantiate the layers, so we would want to do it deliberately.

> Re: allowing the scheduler to run NumPy, pandas, ... code. That's certainly something we should consider. I'm also not aware of any deployments where the scheduler and workers have separate software environments. Do folks who are writing `Layer` classes feel pain around not being able to use libraries like NumPy and pandas?

Yes, there are a number of functions around reshaping, reasoning about nans and dtypes and such that would be useful to be able to use in `Layer` classes. We may be able to get rid of `CallableLazyImport` all together. However, as far as I know we still want to keep the restriction of not deserializing user code/objects on the scheduler, so there are perhaps fewer things that one can do with numpy/pandas than might appear on first blush.

I think that in practice, keeping the layer implementations separate does allow for easier enforcing of serialization restrictions, even if we do want to relax the numpy/pandas restrictions. But if the specific constraints are well-documented and clear, perhaps it's not a big deal.

**Comment by jrbourbeau** [comment]
>  I just want to flag that a consequence of this is that you will need numpy/pandas installed in order to import/instantiate the layers, so we would want to do it deliberately.

Hmm I could be missing something, but I don't quite follow why this would require us to import NumPy/pandas. We would continue to avoid top-level imports in, for example `array/layers.py`, just like we do today in the root-level `layers.py`. We would also add the new `dask.array.layers`, `dask.dataframe.layers` modules to our import check in CI

https://github.com/dask/dask/blob/636b0d664e78febc54434131d6f7f906a6e74f6a/continuous_integration/scripts/test_imports.sh#L17

like we did for `dask.layers` previously

**Comment by rjzamora** [comment]
I assume that we could change the existing behavior, but we currently import a lot if things in `dataframe/__init__` that in-turn import pandas etc.

**Comment by jrbourbeau** [comment]
> I assume that we could change the existing behavior, but we currently import a lot if things in dataframe/__init__ that in-turn import pandas etc.

Ah, I forgot that something like `from dask.dataframe.layers import FooLayer` will run through `dask/dataframe/__init__.py` and not just go straight to `dask/dataframe/layers.py`. In that case, I agree that if we want to keep the NumPy/pandas `import` restrictions then moving layers to a separate `layers/` directory as proposed in https://github.com/dask/dask/issues/7722#issuecomment-850068416 is a better approach.

I suppose the question we need to answer now is whether or not we want to continue to keep the `import` restrictions or not. I'm personally open to relaxing this requirement as it sounds like it would reduce the burden on developers and we're not aware of any cases where relaxing the `import` restriction would be an issue. What do others think?

**Comment by ian-r-rose** [comment]
> EDIT: I do wonder if this will have any implications for a future version of the scheduler which might be written in some other language like C or Go

Good point -- avoiding numpy/pandas would make that somewhat more tractable.

**Comment by jsignell** [comment]
Just to hop on here late :) It seems like the most approachable solution, that leaves our options open re-imports is to have a separate top-level `layers` directory as suggested in https://github.com/dask/dask/issues/7722#issuecomment-850068416. 

In such a directory we could even enforce that pandas and numpy not be imported. 

I am imagining this structure:
```
├── layers
│   ├── array.py
│   ├── core.py
│   ├── dataframe.py
└── └── utils.py
```

**Comment by gjoseph92** [comment]
@jsignell the only issue with that, versus having `array`, `dataframe`, etc. subdirectories like:
```
.
└── layers
    ├── array
    │   ├── core.py
    │   ├── map_overlap.py
    │   ├── optimization.py
    │   ├── rechunk.py
    │   └── slicing.py
    ├── dataframe
    │   ├── core.py
    │   ├── io.py
    │   ├── optimization.py
    │   └── shuffle.py
    ├── blockwise.py
    ├── core.py
    └── optimization.py
```
is that we'd have to collapse a lot of code into one file. For example, `array/overlap.py` and `array/slicing.py` are currently two files (1k and 2k lines respectively); bringing all of that (and more) into one `array.py` file would get really long.

**Comment by gjoseph92** [comment]
> Crazy idea, what if we left these in dask/dataframe or dask/array and just accepted the added requirements on the scheduler. This is somewhat unclean, but we might do it anyway and see if anyone complains. In practice everyone I know deploys the same software environment on the scheduler as they do on the workers. I've always been the one to argue for this cleanliness, but I wanted us to be aware of breaking this cleanliness as an option.

I have been thinking this as well. It's starting to feel like cumbersome and wide-reaching restriction for a relatively niche use case. I'd imagine >95% of users don't care. Additionally, NumPy is already effectively required on the scheduler (it's a transitive dependency of bokeh).

> I do wonder if this will have any implications for a future version of the scheduler which might be written in some other language like C or Go

I agree with this, but only if we actually think we're going to rewrite _all_ of the scheduler in something besides Cython. And beyond that, it would only apply if the high-level graph logic in dask itself were rewritten in a different language too—because that dask code is what lowers HLGs, whether locally with the threaded scheduler or on a cluster with distributed.

If we got to that point, I'd imagine dask graphs were no longer even Python dicts, but some other data structure managed by a compiled library with a Python API, and dask as we know it would be pretty different. If we don't see that happening in a timescale we care about, then even if the whole scheduler were rewritten in Rust by next year, I imagine it would still embed a Python interpreter to basically run `HighLevelGraph.to_dict()`. And in that case, that Python interpreter could still require NumPy and Pandas if it needed to, even if the whole rest of the scheduler was isolated.

All that said—moving some files around and changing imports is not _that_ much work. It doesn't have that many downsides besides a few days' toil. And if we decide we don't like it in the future, it's probably pretty easy to merge things back together. If we think there's a real chance that keeping NumPy and Pandas imports out of the scheduler will make a difference for a meaningful set of users, then it's much easier to adopt that stance now and undo it later than to go the other way around.

**Comment by jsignell** [comment]
Aside from the imports discussion, I  like the idea of having the layers in their own hierarchy. 

As Gabe mentioned there will be multiple files per collection type so I think the options to consider are:

```
.
└── array
    ├── core.py
    ├── map_overlap.py
    ├── optimization.py
└── layers
    ├── array
    │   ├── core.py
    │   ├── map_overlap.py
    │   ├── optimization.py
```

OR

```
.
└── array
    ├── core.py
    ├── map_overlap.py
    ├── optimization.py
    ├── layers
    │   ├── core.py
    │   ├── map_overlap.py
    │   ├── optimization.py
```
